fix: Skip execution control for agent pod. Fixes #12726

jswxstw opened the pull request:
<!-- markdownlint-disable MD041 -- this is rendered within existing HTML, so allow starting without an H1 -->

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

Fixes #12726 

### Motivation

<!-- TODO: Say why you made your changes. -->
Obtain nonexistent node with agent pod in `applyExecutionControl`.

### Modifications

<!-- TODO: Say what changes you made. -->
<!-- TODO: Attach screenshots if you changed the UI. -->
Skip execution control for agent pod.

### Verification

<!-- TODO: Say how you tested your changes. -->
local test

<!--
### Beyond this PR

Thank you for submitting this! Have you ever thought of becoming a Reviewer or Approver on the project?

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable.
See [Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) for more information.

-->

Joibel commented on workflow/controller/exec_control.go:
```diff
@@ -13,9 +13,9 @@ import (
 
 // applyExecutionControl will ensure a pod's execution control annotation is up-to-date
-// kills any pending and running pods when workflow has reached it's deadline
+// kills any pending and running pods(expect agent pod) when workflow has reached its deadline
```
`except` not `expect`.
Also, suggest a space before the `(`

Joibel reviewed (approved):
LGTM.

Looks like the test failures are the common ones which have been happening recently.

jswxstw commented:
> LGTM.
> 
> Looks like the test failures are the common ones which have been happening recently.

Related to: https://github.com/argoproj/argo-workflows/pull/12730#issuecomment-1975644264

agilgur5 commented:
/retest

agilgur5 commented:
> /retest

Hmm that might have failed because the job run is 2 months old. 

Now I clicked the button to update the branch on GH, selected rebase instead of merge, annnd it [broke DCO](https://github.com/argoproj/argo-workflows/pull/12732/checks?check_run_id=24581603842) 😕

agilgur5 reviewed (commented):
I'm wondering if this has a use-case for Agent pods -- the Workflow's deadline should still apply to them.

`woc.wf.NodeId("agent")` seems to work for them in `agent.go`'s `getAgentPodName()` func

jswxstw commented:
> To trigger a re-run I updated the branch via the button on the GH UI, selected rebase instead of merge, annnd it [broke DCO](https://github.com/argoproj/argo-workflows/pull/12732/checks?check_run_id=24581603842)

I fixed it, but there are some new failed tests that appear frequently recently.

> I'm wondering if this has a use-case for Agent pods -- the Workflow's deadline should still apply to them.

I've also thought about it. The workflow's deadline can not apply to agent pods since they do not support being terminated now.

This is somewhat related to another issue I raised, see https://github.com/argoproj/argo-workflows/issues/12692#issuecomment-1960979007.

agilgur5 commented:
/retest

agilgur5 commented:
> I've also thought about it. The workflow's deadline can not apply to agent pods since they do not support being terminated now.
> 
> This is somewhat related to another issue I raised, see [#12692 (comment)](https://github.com/argoproj/argo-workflows/issues/12692#issuecomment-1960979007).

Ah I see, because it doesn't support the `kill` command etc (which is pretty hacky as-is). Could you leave a TODO comment above this line then, e.g. `// TODO: run for Agent Pod as well once it supports termination by the Controller`?

Also when you've thought through some of these situations, it's useful to write them down in the PR description, both for historical context and so that reviewers know without even asking. I use a "### Notes to Reviewers" section in my PRs where there are some changes that require an explanation

Also just wanted to say, I do appreciate all the investigation and work you've done for the Agent and Plugins!

agilgur5 commented:
> I fixed it, but there are some new failed tests that appear frequently recently.

Yea these seem to have been really recent. I'm not sure which PR caused them though 😕

jswxstw commented:
> Ah I see, because it doesn't support the `kill` command etc (which is pretty hacky as-is). Could you leave a TODO comment above this line then, e.g. `// TODO: run for Agent Pod as well once it supports termination by the Controller`?

The agent pod will be deleted once the workflow is completed now, so the pod termination may be unnecessary.

This PR #12441 has support plugin nodes shutdown, but there are a few unresolved issues.
https://github.com/argoproj/argo-workflows/blob/ccb71bdacd13ead899a9005f3078225b5e388715/workflow/controller/operator.go#L1265-L1282
1. Active HTTP nodes should also be shut down here since they also use the agent pod.
2. The workflow's ttl should also apply to all active taskset nodes(HTTP/Plugin) here.

I will submit a new PR to fix it.

agilgur5 commented:
/retest

agilgur5 commented:
> The agent pod will be deleted once the workflow is completed now, so the pod termination may be unnecessary.

For now I suppose. It's a little bit different timing-wise, but close enough. A comment would still be helpful as it is relevant to #12692 / must change if that is implemented

> 1. Active HTTP nodes should also be shut down here since they also use the agent pod.

Agreed

> 2\. The workflow's ttl should also apply to all active taskset nodes(HTTP/Plugin) here.

How is TTL related there? Since it's only deadline related

jswxstw commented:
> How is TTL related there? Since it's only deadline related

Sorry, I made a mistake and I acturally means the workflow's deadline.
I acturally want to say that it is reasonable to skip execution control in `applyExecutionControl` for the agent pod since it will or should be done in `failSuspendedAndPendingNodesAfterDeadlineOrShutdown`.

jswxstw commented:
@agilgur5 Can this PR be merged?

agilgur5 commented:
> A comment would still be helpful as it is relevant to #12692 / must change if that is implemented

This was the only thing I was waiting for. Has actually been on my to-do list to add this myself since I have that permission these days

jswxstw commented:
> This was the only thing I was waiting for. Has actually been on my to-do list to add this myself since I have that permission these days

I updated the PR information, execution control will apply for nodes of the agent pod here now:
https://github.com/argoproj/argo-workflows/blob/735739f16822bb8c869662595a186875b3525307/workflow/controller/operator.go#L312-L315

tooptoop4 commented:
💯

isubasinghe commented:
I guess this should be merged in, I haven't reviewed it personally and will hold off on merging it myself until I have done so. Other approvers are free to merge it.
